gossip: Make gossip listener failure handling more robust

asias opened the pull request:
Since commit 0ed84710d9e, a failure in the gossip listener callbacks or
replication of gossip app state would cause a whole scylla process
to abort.

This is very dangerous if multiple nodes in the cluster have similar
situation and abort at the same time. We recently saw a case where the
host_id of a bootstrapping node was not available in gossip because the
bootstrapping node was removed from gossip since it was not updating
gossip status for some time. But later, the bootstrapping node finished
streaming and advertised NORMAL status. The gossip listener throwed
because the host_id could not be found. As a result, multiple nodes
aborted at the same time.

To improve this, in case of gossip listener failure, instead of abort
immediately, we tried to revert the gossip state to a previous state, so
that the node could learn the new state later and run the listener
again.

Fixes #9657

slivne commented:
@tgrabiec review ping we have a customer that hit this

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
+                if (app == application_state::LOAD ||
+                    app == application_state::CACHE_HITRATES ||
+                    app == application_state::VIEW_BACKLOG) {
+                    continue;
```
This "if" is unsafe in case of extension. We should use a switch without a default case so that compilation fails unless every state is explicitly assigned here.

asias replied:
Well, the switch trick indeed can help extension.

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     // Exceptions thrown from listeners will result in abort because that could leave the node in a bad
     // state indefinitely. Unless the value changes again, we wouldn't retry notifications.
     // Some values are set only once, so listeners would never be re-run.
     // Listeners should decide which failures are non-fatal and swallow them.
```
> Listeners should decide which failures are non-fatal and swallow them.

Why not use this approach?

asias replied:
Because this approach does not work. With this approach, there are two cases

1) no error, happy path

2) there is  failure, the listener either throw (A: terminate whole scylla process) or swallow (B: leave the node in a bad state forever)

This patch introduce 3rd option C for case 2, that is allow gossip to run the listener again to recover. C is better than A and B.

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
+            logger.warn("Started to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+            local_state = old_state;
+            replicate(addr, old_state).get();
```
This can roll back a state change for which we already ran a change listener. In other words, gossip states are no longer monotonic within the process's lifetime. That can break some expectations, and is harder to reason about than monotonic changes. I would rather have us solve this problem in a different way.

asias replied:
First, the failure of the handler should be a very rare event. This patch is about how to handle the worse case scenario. Yes, when we roll back, the listener may have run.  But the inconsistency is supposed to be temporary when the listener is called again. Note, currently, if the failure happens, we abort the whole process while we can probably recover, which is horrible. 

I would prefer a temporary inconsistency over a whole node and even whole cluster abort which is a much much worse. Also, a node can stay inconsistent if it does not get the gossip update temporarily too. 

You are welcome to suggest solution to handle the failure while does not abort the scylla process and not leave the node in an inconsistency state forever.

tgrabiec replied:
> First, the failure of the handler should be a very rare event. This patch is about how to handle the worse case scenario. Yes, when we roll back, the listener may have run. But the inconsistency is supposed to be temporary when the listener is called again. Note, currently, if the failure happens, we abort the whole process while we can probably recover, which is horrible.
> 
> I would prefer a temporary inconsistency over a whole node and even whole cluster abort which is a much much worse. Also, a node can stay inconsistent if it does not get the gossip update temporarily too.

Inconsistency can lead to logic failure in layers above the gossiper, which can lead to worse things than loss of availability. It can lead to loss of data or incorrect query results.

> 
> You are welcome to suggest solution to handle the failure while does not abort the scylla process and not leave the node in an inconsistency state forever.

I don't think there is a general solution. It depends on the case.

In the scenario you mentioned, why does this patch solve the problem? Suppose that some nodes removed the node from gossip, and some not, at the point where it jumps to normal. Those who did, will fail in the listener and, after this patch, revert to the old state. Those who did not, will put the node into the ring. The cluster will be in an inconsistent state.

I also don't understand why we fail on missing host_id, if we don't have any states for the bootstrapping node, why aren't all states, including host id, pulled in?

asias replied:
For example:

n1 is the existing node
n2 is the new bootstrapping node
n2 finishes streaming
n2 advertises NORMAL 
n1 does not receive the NORMAL updates yet for network issues
n1 removes n2 from gossip because n2 does not update gossip, fat_client  removal
n1 receives NORMAL update but all other state of n2 is removed
n1 throws in the handler because it can not find the host_id

asias replied:
This patch helps the problem because after n1 rolls back, n1 has a chance to talk with n2 and get all the states from n2 again and run the normal state handler successfully.

asias replied:
Again, the inconsistency is temporary. The general statement inconsistency can cause data inconsistency is true. But this can happen if a node is partitioned and does not know the new node is in NORMAL status too. This patch is not making it worse.

asias replied:
Gossip verbs are one way, so 

> I'm still interested in [#9658 (comment)](https://github.com/scylladb/scylla/pull/9658#discussion_r765690600), which may be the culprit of the issue which motivated this. But this PR can be considered separately.

Yes, this patch handles more generic problems: what happens when a gossip handler fails. The discussion there is just one of the reasons how the handler can fail.

asias replied:
Gossip messages are async and one way message. Removal can not be mutually exclusive with what you called syns. Because we do not know when the diff would come and can not wait for it forever which will block gossip runs. 
```
1) 
take the lock
find the gossip state diff and ask the peer to send diff
the diff never arrives

2) 
can not take the lock to remove a node
```

tgrabiec replied:
Is it possible to fix with the current protocol?

asias replied:
This will change the verbs which will introduce incompatibility which is a much bigger issue.

tgrabiec replied:
I opened https://github.com/scylladb/scylla/issues/9803 for this

tgrabiec commented:
We could avoid aborts and keep the state monotonic. Instead of reverting state on listener failure, we keep retrying to execute the listener in the background, not blocking gossip activity for other endpoints. All activity for this endpoint will be serialized using the existing semaphore, so this node's state will not advance until the failed listener eventually succeeds. 

In the scenario you mentioned, admin intervention will be required to remedy the situation (e.g. rolling restart), because the listener will never succeed. But the cluster won't crash.

asias commented:
> We could avoid aborts and keep the state monotonic. Instead of reverting state on listener failure, we keep retrying to execute the listener in the background, not blocking gossip activity for other endpoints. All activity for this endpoint will be serialized using the existing semaphore, so this node's state will not advance until the failed listener eventually succeeds.
> 
> In the scenario you mentioned, admin intervention will be required to remedy the situation (e.g. rolling restart), because the listener will never succeed. But the cluster won't crash.

Since the listener takes the semaphore which will block other updates to this same node, retrying will not recover. Rolling back solution can recover the issue automatically, has the downside of moving state back and partial listener execution. 

However, on the other hand, the impact is limited to the node in question only. Perhaps this is good enough in practice. It is much better than abort the node already. 

Taking the sem for the node in question could  block the whole gossip loop (e.g., the node decide to evict from gossip) or prevent other fiber not to finish forever. 

I think we can take another approach to isolate the node. 

1. the listener is not retried in case of error, it does not throw, so we do not abort the whole node. Retry will probably not help anyway. 
2. mark the node as isolated
3. do not process any messages from the isolated node
4. do not apply any change for the isolated node
5. mark the isolated node as down
6. tell the user we need a rolling restart to recover

asias commented:
I had something like below. The implementation is pretty simple.

```
diff --git a/gms/gossiper.hh b/gms/gossiper.hh
index 5bef4c1719..7f25f964ef 100644
--- a/gms/gossiper.hh
+++ b/gms/gossiper.hh
@@ -250,6 +250,10 @@ class gossiper : public seastar::async_sharded_service<gossiper>, public seastar
     // Replicates given value to all other shards.
     // The value must be kept alive until completes and not change.
     future<> replicate(inet_address, application_state key, const versioned_value& value);
+
+    std::unordered_set<gms::inet_address> _isolated_nodes_due_to_listener_failure;
+
+    bool is_isolated(const gms::inet_address& addr);
 public:
     explicit gossiper(abort_source& as, feature_service& features, const locator::shared_token_metadata& stm, netw::messaging_service& ms, db::config& cfg, gossip_config gcfg);
 
diff --git a/gms/gossiper.cc b/gms/gossiper.cc
index 738a3dbcbd..2244983ffc 100644
--- a/gms/gossiper.cc
+++ b/gms/gossiper.cc
@@ -202,6 +202,10 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
         return make_ready_future<>();
     }
 
+    if (is_isolated(from.addr)) {
+        return make_ready_future<>();
+    }
+
     /* If the message is from a different cluster throw it away. */
     if (syn_msg.cluster_id() != get_cluster_name()) {
         logger.warn("ClusterName mismatch from {} {}!={}", from.addr, syn_msg.cluster_id(), get_cluster_name());
@@ -296,6 +300,10 @@ future<> gossiper::handle_ack_msg(msg_addr id, gossip_digest_ack ack_msg) {
         return make_ready_future<>();
     }
 
+    if (is_isolated(id.addr)) {
+        return make_ready_future<>();
+    }
+
     msg_proc_guard mp(*this);
 
     auto g_digest_list = ack_msg.get_gossip_digest_list();
@@ -392,6 +400,10 @@ future<> gossiper::handle_ack2_msg(msg_addr from, gossip_digest_ack2 msg) {
         return make_ready_future<>();
     }
 
+    if (is_isolated(from.addr)) {
+        return make_ready_future<>();
+    }
+
     msg_proc_guard mp(*this);
 
     auto& remote_ep_state_map = msg.get_endpoint_state_map();
@@ -400,6 +412,9 @@ future<> gossiper::handle_ack2_msg(msg_addr from, gossip_digest_ack2 msg) {
 }
 
 future<> gossiper::handle_echo_msg(gms::inet_address from, std::optional<int64_t> generation_number_opt) {
+    if (is_isolated(from)) {
+        return make_exception_future(std::runtime_error(format("Node {} is isolated by {}", from, get_broadcast_address())));
+    }
     bool respond = true;
     if (!_advertise_myself) {
         respond = false;
@@ -1507,6 +1522,12 @@ void gossiper::mark_alive(inet_address addr, endpoint_state& local_state) {
     //     real_mark_alive(addr, local_state);
     //     return;
     // }
+
+    if (is_isolated(addr)) {
+        logger.warn("Skipped to mark node {} as alive because it is isloated", addr);
+        return;
+    }
+
     auto inserted = _pending_mark_alive_endpoints.insert(addr).second;
     if (inserted) {
         // The node is not in the _pending_mark_alive_endpoints
@@ -1687,6 +1708,10 @@ bool gossiper::is_silent_shutdown_state(const endpoint_state& ep_state) const{
     return false;
 }
 
+bool gossiper::is_isolated(const gms::inet_address& node) {
+    return _isolated_nodes_due_to_listener_failure.contains(node);
+}
+
 // Runs inside seastar::async context
 void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state, const endpoint_state& remote_state) {
     // don't assert here, since if the node restarts the version will go back to zero
@@ -1701,13 +1726,42 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     utils::chunked_vector<application_state> changed;
     auto&& remote_map = remote_state.get_application_state_map();
 
+    bool replicate_failed = false;
+    bool listener_failed = false;
+
+    auto isolate_node = [&] {
+        try {
+            logger.warn("Node {} is isolated by gossip because it failed to run gossip listener. A rolling restart is needed to recover!", addr);
+            _isolated_nodes_due_to_listener_failure.insert(addr);
+            container().invoke_on_others([addr] (gossiper& g) {
+                g._isolated_nodes_due_to_listener_failure.insert(addr);
+            }).get();
+            logger.info("Mark node {} as DOWN because it is isloated", addr);
+            convict(addr);
+        } catch (...) {
+            logger.error("Failed to isolate node {}", addr, std::current_exception());
+        }
+    };
+
     // Exceptions thrown from listeners will result in abort because that could leave the node in a bad
     // state indefinitely. Unless the value changes again, we wouldn't retry notifications.
     // Some values are set only once, so listeners would never be re-run.
     // Listeners should decide which failures are non-fatal and swallow them.
     auto run_listeners = seastar::defer([&] () noexcept {
-        for (auto&& key : changed) {
-            do_on_change_notifications(addr, key, remote_map.at(key));
+        if (replicate_failed) {
+            logger.warn("Skipped to run gossip listener for node={}, changed={}, remote_map={}",
+                    addr, changed, remote_map);
+            return;
+        }
+        try {
+            for (auto&& key : changed) {
+                do_on_change_notifications(addr, key, remote_map.at(key));
+            }
+        } catch (...) {
+            listener_failed = true;
+            logger.warn("Failed to run gossip listener for node={}, changed={}, remote_map={}: {}",
+                    addr, changed, remote_map, std::current_exception());
+            isolate_node();
         }
     });
 
@@ -1716,7 +1770,14 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     // would be inconsistent across shards. Changes listeners depend on state
     // being replicated to all shards.
     auto replicate_changes = seastar::defer([&] () noexcept {
-        replicate(addr, remote_map, changed).get();
+        try {
+            replicate(addr, remote_map, changed).get();
+        } catch (...) {
+            replicate_failed = true;
+            logger.warn("Failed to replicate app state for node={}, changed={}, remote_map={}: {}",
+                    addr, changed, remote_map, std::current_exception());
+            isolate_node();
+        }
     });
 
     // we need to make two loops here, one to apply, then another to notify,
```

asias commented:
Changes in v2 (cc9cdaec83b99195c02d07e6f0a44b53c12802a5):
- Do not roll back gossip state
- Isolate the node in question instead

asias commented:
Ping for merge. @avikivity @tgrabiec

tgrabiec commented:
I'm still interested in https://github.com/scylladb/scylla/pull/9658#discussion_r765690600, which may be the culprit of the issue which motivated this. But this PR can be considered separately.

tgrabiec commented:
Blocking gossip is controversial since we still rely on `ring_delay` for correctness of topology changes. So blocking gossip indefinitely can increase the risk of violating data consistency.

asias commented:
> Blocking gossip is controversial since we still rely on `ring_delay` for correctness of topology changes. So blocking gossip indefinitely can increase the risk of violating data consistency.

This patch does not block gossip. See the code. It only isolates the problematic node in question.

asias commented:
> I'm still interested in [#9658 (comment)](https://github.com/scylladb/scylla/pull/9658#discussion_r765690600), which may be the culprit of the issue which motivated this. But this PR can be considered separately.

Yes, this patch handles more generic problems: what happens when a gossip handler fails. The discussion there is just one of the reasons how the handler can fail.

tgrabiec commented:
> > Blocking gossip is controversial since we still rely on `ring_delay` for correctness of topology changes. So blocking gossip indefinitely can increase the risk of violating data consistency.
> 
> This patch does not block gossip. See the code. It only isolates the problematic node in question.

I meant blocking gossip messages about the state of one of the nodes.

tgrabiec commented:
> Yes, this patch handles more generic problems: what happens when a gossip handler fails. The discussion there is just one of the reasons how the handler can fail.

We may want this change, but still, if we can prevent listeners from failing, it's better. Isolating the node is still a bad effect.

asias commented:
> > > Blocking gossip is controversial since we still rely on `ring_delay` for correctness of topology changes. So blocking gossip indefinitely can increase the risk of violating data consistency.
> > 
> > 
> > This patch does not block gossip. See the code. It only isolates the problematic node in question.
> 
> I meant blocking gossip messages about the state of one of the nodes.

At the same time, the blocked node is being marked as DOWN by existing nodes. The blocked node itself will find other existing nodes are DOWN. I think this is pretty safe.

asias commented:
> > Yes, this patch handles more generic problems: what happens when a gossip handler fails. The discussion there is just one of the reasons how the handler can fail.
> 
> We may want this change, but still, if we can prevent listeners from failing, it's better. Isolating the node is still a bad effect.

Not having this protection to prevent abort a node which could abort the whole cluster is just a bomb handling there. Preventing is one thing. We can not guarantee all listeners do not fail in all cases.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1701,13 +1729,42 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
+            convict(addr);
+        } catch (...) {
+            logger.error("Failed to isolate node {}", addr, std::current_exception());
+        }
```
Who are we isolating here? The node we got the update for? But it didn't have any problem.  We should isolate ourselves!

A different approach to the problem is to make sure we never ever fail. That requires pre-allocating state for each node when we first add it, and it's not simple.

asias replied:
Consider A sends updates to B. B fails to call gossip listeners for A. It could that A has something wrong and sends B something unexpected. It is not necessarily B fails to apply just because memory allocation. 

Currently, B will abort itself. After this patch, B does not abort. Instead, B stops everything with A and marks A as DOWN. Fom B's point of view, node A is down. This has minimum limited impact. The impact is limited with the node A. 

Even if it is B fails to call the listener because of memory allocation. It is the fault of node B, it does not matter. It does not change the fact B failed the listener for A. B either stop everything with A or abort itself,  to prevent B could possibly have a wrong state for A forever.

avikivity replied:
Let's say the cause was bad_alloc. In the current code, the node restarts and will recover. With the new code, A thinks it updated B, but B is in some mixed state with some listeners updated and the others not updated.

Maybe we survive longer, but the total number of states we have to consider is much larger, and so the potential for encountering new bugs is larger.

asias replied:
A can not assume anything about B even if A sends something to B with / without this patch.  After this patch, A will think B is down because B does not respond anything with A. B will mark A as DOWN too. The state that A and B mark each other DOWN looks pretty safe, as if they were isolated from network. 

I agree it makes things a bit more complicated, but it can prevent possible ALL nodes down situation which is a total disaster. 

Even if with this patch, we need to notify the user to perform rolling restart to recover the cluster to a health state. The problem of having more total number of states will be gone after rolling restart. In the end, we will have similar results as we abort immediate except there is no abort and no whole cluster down.

avikivity replied:
The problem here is that the number of cases explodes. First we have a failure during gossip update. It's already a rare occurrence and we have no testing for it. And now the rare case survives and can accumulate more rare cases.

I don't agree about the whole cluster going down. If it's a transient bug (memory allocation failure) then there's no reason to expect other nodes will hit it. If it's a software bug, then all nodes will have it and all nodes will start isolating each other.

One exception is during upgrade, if the first upgraded node goes crazy and causes exceptions in all the old nodes, then isolating it is a good choice. But it only works for gossip, if it generates fatal exceptions in other components it doesn't help.

asias replied:
> I don't agree about the whole cluster going down. If it's a transient bug (memory allocation failure) then there's no reason to expect other nodes will hit it. If it's a software bug, then all nodes will have it and all nodes will start isolating each other.

Your bad_alloc example will mostly certainly not trigger whole cluster down. However, we saw whole cluster down happened in practice with our user! This is the motivation of this patch as part of the whole fix (fix in generic gossip layer and fix in specific user of gossip) in the first place. It is not that all nodes will start isolating each other. The isolating happens between only existing nodes with the problematic node. 

In the long run, we are moving away more critical usages (e.g., membership ) from gossip, leaving uncritical things (e.g., cache hit rate, loads) only. It is becoming more safe in terms of correctness even when gossip has more cases you worried.

avikivity replied:
I'm still worried that this moves us into untested code. We don't even have a test to exercise this once, and it can happen in many places.

An abort() is bad, yes, but running an untested configuration is also bad.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1701,13 +1729,42 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
+            logger.info("Mark node {} as DOWN because it is isloated", addr);
+            convict(addr);
+        } catch (...) {
+            logger.error("Failed to isolate node {}", addr, std::current_exception());
```
If we failed to apply a gossip update, we're likely to fail here too.

tgrabiec replied:
Not necessarily, the failure can be due to listener-specific logic. That's the case which motivated this patch. See https://github.com/scylladb/scylla/pull/9658#discussion_r765358252

asias replied:
Yes.

asias commented:
@avikivity  Ping.

asias commented:
Changes in v3 (4444f62fadc0570e7af9371608a7013da0a956ac):
- Skip sending echo in failure_detector_loop_for_node explicitly if a node is isolated.
